halo2-base/src/safe_types/mod.rs

```diff
@@ -0,0 +1,133 @@
+pub use crate::{
+    gates::{
+        flex_gate::GateInstructions,
+        range::{RangeChip, RangeInstructions},
+    },
+    utils::ScalarField,
+    AssignedValue, Context,
+    QuantumCell::{self, Constant, Existing, Witness},
+};
+use std::cmp::{max, min};
+
+type RawAssignedValues<F> = Vec<AssignedValue<F>>;
+
+const BITS_PER_BYTE: usize = 8;
+// Each AssignedValue can at most represent 8 bytes.
+const MAX_BYTE_PER_ELEMENT: usize = 8;
+
+// SafeType's goal is to avoid out-of-range undefined behavior.
+// When building circuits, it's common to use mulitple AssignedValue<F> to represent
+// a logical varaible. For example, we might want to represent a hash with 32 AssignedValue<F>
+// where each AssignedValue represents 1 byte. However, the range of AssignedValue<F> is much
+// larger than 1 byte(0~255). If a circuit takes 32 AssignedValue<F> as inputs and some of them
+// are actually greater than 255, there could be some undefined behaviors.
+// SafeType gurantees the value range of its owned AssignedValue<F>. So circuits don't need to
+// do any extra value checking if they take SafeType as inputs.
+#[derive(Clone, Debug)]
+pub struct SafeType<F: ScalarField, const B: usize, const L: usize> {
+    value: RawAssignedValues<F>,
+}
+
+impl<F: ScalarField, const B: usize, const L: usize> SafeType<F, B, L> {
+    // new is private so Safetype can only be constructed by this crate.
+    fn new(raw_values: RawAssignedValues<F>) -> Self {
+        Self { value: raw_values }
+    }
+
+    // Return values in littile-endian.
+    pub fn value(&self) -> &RawAssignedValues<F> {
+        &self.value
+    }
+
+    // length of value() must equal to value_length().
+    pub fn value_length() -> usize {
+        L
+    }
+
+    // All elements in value() need to be in [0, element_limit()].
+    pub fn element_limit() -> u64 {
+        ((1u128 << B) - 1) as u64
+    }
+
+    // Each element in value() has element_bits() bits.
+    pub fn element_bits() -> usize {
+        B
+    }
+}
+
+pub type SafeBool<F> = SafeType<F, 1, 1>;
+pub type SafeUint8<F> = SafeType<F, 8, 1>;
+pub type SafeUint16<F> = SafeType<F, 16, 1>;
+pub type SafeUint32<F> = SafeType<F, 32, 1>;
+pub type SafeUint64<F> = SafeType<F, 64, 1>;
+pub type SafeUint128<F> = SafeType<F, 64, 2>;
+pub type SafeUint256<F> = SafeType<F, 64, 4>;
+pub type SafeBytes32<F> = SafeType<F, 8, 32>;
+
+pub struct SafeTypeChip<F: ScalarField> {
```

Comment:
We don't want to call any function in `RangeInstructions` in `SafeTypeChip` because we have different input types, what's the point to extend  `RangeInstructions`?

Reply:
I mean right now to create a `SafeTypeChip` you must move a `RangeChip` and own it. I think the more common use case is that you already have a `RangeChip` used elsewhere in your circuit code, and at some points you want to be able to call `range.<SafeTypeChips function>`. That's what I meant by `trait` instead of `struct` for `SafeTypeChip`. 

`SafeTypeChip` uses `range_check`, which is a function in `RangeInstructions`, which is why I said to extend `RangeInstructions`.

Reply:
In this case I think shared ownership makes more sense. Use Arc<RangeChip> now

Reply:
OK let's move it to `RangeChip` then. Somehow this is semantically correct - it is indeed related to range.